app/lib/frontend/email_sender.dart

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:logging/logging.dart';
+import 'package:mailer/mailer.dart';
+import 'package:mailer/smtp_server.dart';
+
+import '../shared/email.dart';
+import 'models.dart' show PrivateKey;
+
+final _logger = new Logger('pub.email');
+
+/// Sets the active [EmailSender].
+void registerEmailSender(EmailSender value) =>
+    ss.register(#_email_sender, value);
+
+/// The active [EmailSender].
+EmailSender get emailSender => ss.lookup(#_email_sender) as EmailSender;
+
+/// Sends emails using credentials periodically fetched from Datastore.
+class EmailSender {
+  final DatastoreDB _db;
+  SmtpServer _server;
+  DateTime _lastUpdated;
+  EmailSender(this._db);
+
+  /// Sends a [message] and returns when the operation completed.
+  /// Errors are only logged, they do not block the processing.
+  Future sendMessage(EmailMessage message) async {
+    await _update();
+    final logText = '(${message.subject}) '
+        'from ${message.from} '
+        'to ${message.recipients.join(', ')}';
+    if (_server == null) {
+      _logger.info('Not sending e-mail (SMTP not configured): $logText.');
+    } else {
+      _logger.info('Sending e-mail: $logText...');
+      try {
+        final result = await send(_toMessage(message), _server);
+        for (var sendReport in result) {
+          sendReport?.validationProblems?.forEach((p) {
+            _logger.info('E-mail problem: ${p.code} ${p.msg}.');
+          });
+        }
+      } catch (e, st) {
+        _logger.severe('Sending e-mail failed: $logText.', e, st);
+      }
+    }
+  }
+
+  Future _update() async {
+    final now = new DateTime.now().toUtc();
+    if (_lastUpdated != null && now.difference(_lastUpdated).inMinutes < 10) {
+      return;
+    }
+    final list = await _db.lookup([
+      _db.emptyKey.append(PrivateKey, id: 'smtp.username'),
```

Comment:
Please add a few comments documenting this... maybe add a configuration entry in the `README.md` that explains how to set the smtp secrets...

Reply:
Also either add a new kind for storing this kind of data... or document that `PrivateKey` is used to stored secrets injected into the application.. ie. in `models.dart` we should have a doc comment on `PrivateKey` saying what it is used for..

We could also remove the `PrivateKey` kind, and add a `Secret` kind instead... Just because `PrivateKey` is rather misleading.

Reply:
I'll add a `Secret` in this PR, and remove `PrivateKey` in a separate one (I want to make sure that we don't need it anymore).